components/fxa-client/ios/FxAClient/FxAccountManager.swift

```diff
@@ -0,0 +1,584 @@
+            self.processEvent(event: .authenticationError) {}
+        }
+    }
+
+    internal func requireAccount() -> FirefoxAccount {
+        if let acct = account {
+            return acct
+        }
+        preconditionFailure("initialize() must be called first.")
+    }
+
+    internal func requireConstellation() -> DeviceConstellation {
+        if let cstl = constellation {
+            return cstl
+        }
+        preconditionFailure("account must be set (sets constellation).")
+    }
+
+    // swiftlint:enable function_body_length
+}
+
+// swiftlint:enable type_body_length
+
+class FxAStatePersistenceCallback: PersistCallback {
+    weak var manager: FxaAccountManager?
+
+    public init(manager: FxaAccountManager) {
+        self.manager = manager
+    }
+
+    func persist(json: String) {
+        manager?.accountStorage.write(json)
+    }
+}
+
+/**
+ * States of the [FxaAccountManager].
+ */
+internal enum AccountState {
+    case start
+    case notAuthenticated
+    case authenticationProblem
+    case authenticatedNoProfile
+    case authenticatedWithProfile
+}
+
+/**
+ * Base class for [FxaAccountManager] state machine events.
+ * Events aren't a simple enum class because we might want to pass data along with some of the events.
+ */
+internal enum Event {
+    case initialize
+    case accountNotFound
+    case accountRestored
+    case authenticated(authData: FxaAuthData)
+    case authenticationError /* (error: AuthException) */
+    case recoveredFromAuthenticationProblem
+    case fetchProfile
+    case fetchedProfile
+    case failedToFetchProfile
+    case logout
+}
+
+public enum FxaAuthType {
+    case existingAccount
+    case signin
+    case signup
+    case pairing
+    case recovered
+    case other(reason: String)
+
+    internal static func fromActionQueryParam(_ action: String) -> FxaAuthType {
+        switch action {
+        case "signin": return .signin
+        case "signup": return .signup
+        case "pairing": return .pairing
+        default: return .other(reason: action)
+        }
+    }
+}
+
+public struct FxaAuthData {
+    public let code: String
+    public let state: String
+    public let authType: FxaAuthType
+
+    public init(code: String, state: String, actionQueryParam: String) {
+        self.code = code
+        self.state = state
+        authType = FxaAuthType.fromActionQueryParam(actionQueryParam)
+    }
+}
+
+public struct DeviceConfig {
+    let name: String
+    let type: DeviceType
+    let capabilities: [DeviceCapability]
+
+    public init(name: String, type: DeviceType, capabilities: [DeviceCapability]) {
+        self.name = name
+        self.type = type
+        self.capabilities = capabilities
+    }
+}
+
+public enum Scope {
+    // Necessary to fetch a profile.
+    public static let profile: String = "profile"
+    // Necessary to obtain sync keys.
+    public static let sync: String = "https://identity.mozilla.com/apps/oldsync"
```

Comment:
nit: please name this `oldsync` for consistency with how we name it in other places